File: Data Structures & Algorithms/remove-element/submission-3.swift
class Solution {
    func removeElement(_ nums: inout [Int], _ val: Int) -> Int {
        var left = 0
        var right = nums.count - 1
        while left <= right {
            if nums[left] == val {
                nums[left] = nums[right]
                right -= 1
            } else {
                left += 1
            }
        }
        return right + 1
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-3.kt
class Solution {
    fun trap(height: IntArray): Int {
        var leftMax = height[0]
        var rightMax = height[height.count() - 1]
        var left = 1
        var right = height.count() - 2
        var water = 0
        while (left <= right) {
            if (leftMax <= rightMax) {
                water += maxOf(minOf(leftMax, rightMax) - height[left], 0)
                leftMax = maxOf(leftMax, height[left])
                left += 1
            } else {
                water += maxOf(minOf(leftMax, rightMax) - height[right], 0)
                rightMax = maxOf(rightMax, height[right])
                right -= 1
            }
        }
        return water
    }
}
